capsules/src/console.rs

```diff
@@ -179,7 +167,7 @@ impl<U: UART> Console<'a, U> {
                     app.read_len = read_len;
                     self.rx_buffer.take().map(|buffer| {
                         self.rx_in_progress.set(app_id);
-                        self.uart.receive(buffer, app.read_len);
+                        let (_err, _opt) = self.uart.receive_buffer(buffer, app.read_len);
```

Comment:
First, I think you should take the  return code from `receive buffer` into account here and not ignore it.

Second, I don't like how the method is written and I propose the following:
```rust
    fn receive_new(&self, app_id: AppId, app: &mut App, len: usize) -> ReturnCode {
        let rx_buf = match self.rx_buffer.take() {
            Some(buf) => buf,
            None => return ReturnCode::EINVAL,
        };

        match app.read_buffer {
            Some(ref slice) => {
                let read_len = cmp::min(len, slice.len());
                if read_len > rx_buf.len() {
                    // For simplicity, impose a small maximum receive length
                    // instead of doing incremental reads
                    ReturnCode::EINVAL
                } else {
                    app.read_len = read_len;
                    self.rx_in_progress.set(app_id);
                    let (err, _opt) = self.uart.receive_buffer(rx_buf, app.read_len);
                    err
                }
            }
            None => {
                // Must supply read buffer before performing receive operation
                ReturnCode::EINVAL
            }
        }
    }

```

Reply:
I agree that it's better to not ignore the error; the implementation currently does because the prior one did, and I was trying to do the simplest port possible without rewriting everything. For example, if you handle the error, then you must handle the option: otherwise you have a memory leak on an error. E.g., in your code, since rx_buffer is taken, and passed to receive_buffer, but then ignored on the return, it'll be lost. I think that if there's an error, you need to put it back into self.rx_buffer.

That being said, if people are on-board with the HIL and its implications, then I agree, going through all of the code to properly handle error cases is important. Of course, this will make the PR even bigger, if every function is rewritten...

Reply:
Okay, I will check the HIL instead!

Regarding `memory leak` do you mean that we lose the reference to the `static mut buffer`? 
Well, technically I don't think it is memory leak because we can reclaim the buffer `console::READ_BUF` but it is unsafe ofc and not intended!

Reply:
Yes, that's what I mean. 

But the fact that it's initialized from console::READ_BUF is due to the boot sequence; the implementation of these functions doesn't require it's from console::READ_BUF. I.e., we should not assume something in main.rs when considering if an implementation within the module has a problem.